File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/message/ReadVariableAttributeResponse.java
package org.opcfoundation.webserver.types.message;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.sdk.core.ValueRank;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.jspecify.annotations.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaDataType;

public class ReadVariableAttributeResponse {
    final LocalizedText displayName;
    final LocalizedText description;
    final NodeId dataTypeId;
    final Integer valueRank;
    final Integer accessLevel;
    final Boolean historizing;

    public ReadVariableAttributeResponse(
            LocalizedText displayName,
            LocalizedText description,
            UaDataType dataType,
            Boolean writable,
            Boolean historizing,
            @Nullable Integer valueRank)
    {
        this.displayName = displayName;
        this.description = description;
        this.dataTypeId = dataType.nodeId();
        this.valueRank = (null == valueRank) ? ValueRank.Scalar.getValue() : valueRank;
        this.accessLevel = (writable) ? AccessLevel.CurrentWrite.getValue() | AccessLevel.CurrentRead.getValue() : AccessLevel.CurrentRead.getValue();
        this.historizing = historizing;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public LocalizedText getDescription()
    {
        return description;
    }

    public NodeId getDataTypeId() {
        return dataTypeId;
    }

    public Integer getValueRank() {
        return valueRank;
    }

    public Integer getAccessLevel() {
        return accessLevel;
    }

    public Boolean getHistorizing() {
        return historizing;
    }
}

File: opcua-webservice-node/src/server/service/message/reactiveobject/ReadVariableAttributeResponse.ts
import { UaAccessLevel, UaLocalizedText, UaNodeId, UaValueRank } from "opcua-webapi-ts";
import { UaDataType } from "../../../addressspace/nodes/UaDataType";

export class ReadVariableAttributeResponse {
    private readonly _displayName: UaLocalizedText;
    private readonly _description: UaLocalizedText;
    private readonly _dataTypeId: UaNodeId;
    private readonly _valueRank: number;
    private readonly _accessLevel: number;
    private readonly _historizing: boolean;

    constructor(
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        dataType: UaDataType,
        writable: boolean,
        historizing: boolean,
        valueRank: number | null,
    ) {
        this._displayName = displayName;
        this._description = description;
        this._dataTypeId = dataType.nodeId;
        this._valueRank = valueRank ?? UaValueRank.Scalar;
        this._accessLevel = writable
            ? UaAccessLevel.CurrentWrite | UaAccessLevel.CurrentRead
            : UaAccessLevel.CurrentRead;
        this._historizing = historizing;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get description(): UaLocalizedText {
        return this._description;
    }

    get dataTypeId(): UaNodeId {
        return this._dataTypeId;
    }

    get valueRank(): number {
        return this._valueRank;
    }

    get accessLevel(): number {
        return this._accessLevel;
    }

    get historizing(): boolean {
        return this._historizing;
    }
}